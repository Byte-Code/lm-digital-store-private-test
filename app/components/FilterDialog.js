import React, { Component, PropTypes } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import styled from 'styled-components';
import { List, Set } from 'immutable';
import RemoveIcon from 'material-ui/svg-icons/content/remove-circle-outline';
import UndoIcon from 'material-ui/svg-icons/content/undo';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  color: #fff;
`;

const Header = styled.div`
  display: flex;
  height: 41px;
  justify-content: space-between;
  margin-bottom: 30px;
`;

const Button = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  &>p {
    color: #fff;
    font-family: ${props => props.fFamily};
    font-size: ${props => props.fSize};
    text-decoration: ${props => props.tDeco};
    margin-left: 12px;
  }
`;

const GroupWrapper = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  margin-bottom: 10px;
`;

const GroupTitle = styled.div`
  font-size: 16px;
  text-transform: uppercase;
  margin-bottom: 13px;
  font-family: LeroyMerlinSans Light;
`;

const FilterWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  &>div {
    margin-right: 10px;
  }
  &>div:last-child {
    margin-right: 0;
  }
`;

const Filter = styled.div`
  width: 150px;
  height: 42px;
  border-radius: 20px;
  margin-bottom: 20px;
  background-color: ${props => (props.isActive ? '#67cb33' : '#efefef')};
  color: ${props => (props.isActive ? '#efefef' : '#333333')}
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  &>p {
    line-height: 20px;
    font-size: 16px;
    text-align: center;
  }
`;

function getAllProducts(filterGroups) {
  filterGroups.map(g => (g.get('filters').reduce((acc, f) => (acc.push(f.get('products'))), List())));
}

function filterProducts(filterGroups, activeFilters) {
  return filterGroups.map(g => {
    return g.get('filters').reduce((acc, f) => {
      if (activeFilters.includes(f.get('code'))) {
        return acc.push(f.get('products'));
      }
      return acc;
    }, List()).flatten().toSet();
  }).filterNot(f => f.isEmpty()).toArray();
}

export default class FilterDialog extends Component {
  static propTypes = {
    filterGroups: ImmutablePropTypes.list.isRequired,
    activeFilters: ImmutablePropTypes.list.isRequired,
    handleClose: PropTypes.func.isRequired,
    resetFilters: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props);

    this.state = {
      active: props.activeFilters,
      totalProducts: []
    };
  }

  getTotalProducts() {
    const { filterGroups } = this.props;
    const { active } = this.state;
    if (active.isEmpty()) {
      return getAllProducts(filterGroups);
    }
    console.log();
    return Set.intersect(filterProducts(filterGroups, active));
    // this.setState({ totalProducts });
  }

  renderFilterGroups() {
    const { filterGroups } = this.props;
    return filterGroups.map(g => (
      <GroupWrapper key={g.get('code')}>
        <GroupTitle>{g.get('group')}</GroupTitle>
        <FilterWrapper>
          {g.get('filters').map(f => (
            <Filter key={f.get('code')}>
              <p>{f.get('name')}</p>
            </Filter>
          ))}
        </FilterWrapper>
      </GroupWrapper>
    ));
  }

  render() {
    const { handleClose, resetFilters } = this.props;
    // this.getTotalProducts();
    console.log(this.getTotalProducts());

    return (
      <Wrapper>
        <Header>
          <Button fFamily="LeoryMerlinSans Light" fSize="20px" onClick={handleClose}>
            <RemoveIcon color="#fff" style={{ height: 25, width: 25 }} />
            <p>Chiudi Filtri</p>
          </Button>
          <Button tDeco="underline" fSize="16px" onClick={resetFilters}>
            <UndoIcon color="#fff" style={{ height: 25, width: 25 }} />
            <p>Reset Filtri</p>
          </Button>
        </Header>
        {this.renderFilterGroups()}
      </Wrapper>
    );
  }
}
