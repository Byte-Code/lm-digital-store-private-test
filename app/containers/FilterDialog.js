import { connect } from 'react-redux';

import * as filtersActions from '../actions/filtersActions';
import { getItemCount, getTempAvailability, getTempFilters } from '../reducers/selectors';
import FilterDialog from '../components/FilterDialog';

const mapStateToProps = (state, ownProps) => ({
  availability: getTempAvailability(state),
  activeFilters: getTempFilters(state),
  itemCount: getItemCount(state, ownProps.categoryCode)
});

const mapDispatchToProps = {
  ...filtersActions
};

export default connect(mapStateToProps, mapDispatchToProps)(FilterDialog);
