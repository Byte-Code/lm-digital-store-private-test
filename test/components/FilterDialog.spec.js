import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';
import Toggle from 'material-ui/Toggle';

import FilterDialog, { Filter } from '../../app/components/FilterDialog';

const applyTempFilters = jest.fn();
const handleClose = jest.fn();
const initTempFilters = jest.fn();
const resetTempFilters = jest.fn();
const toggleTempFilter = jest.fn();
const toggleTempAvailability = jest.fn();
const filterGroups = fromJS([
  {
    code: 'group0',
    filters: [
      { code: 'filter0', products: [0, 1, 2] },
      { code: 'filter1', products: [3, 4, 5] },
      { code: 'filter2', products: [0, 1, 2] }
    ]
  },
  {
    code: 'group1',
    filters: [
      { code: 'filter3', products: [0, 1, 2] },
      { code: 'filter4', products: [3, 4, 9] },
      { code: 'filter5', products: [6, 7, 8] }
    ]
  }
]);
const activeFilters = fromJS(['filter1']);
const result = shallow(
  <FilterDialog
    filterGroups={filterGroups}
    handleClose={handleClose}
    categoryCode="CAT123"
    applyTempFilters={applyTempFilters}
    initTempFilters={initTempFilters}
    toggleTempFilter={toggleTempFilter}
    toggleTempAvailability={toggleTempAvailability}
    resetTempFilters={resetTempFilters}
    activeFilters={activeFilters}
  />
);

describe('FilterDialog', () => {
  it('should call handleClose when the button is clicked', () => {
    result.find('#close-filterDialog').simulate('click');
    expect(handleClose).toHaveBeenCalled();
  });

  it('should call toggleTempFilter when a filter is clicked', () => {
    result.find(Filter).at(0).simulate('click');
    expect(toggleTempFilter).toHaveBeenCalledWith('filter0');
    expect(result).toMatchSnapshot();
  });

  it('should call toggleTempAvailability availability is toggled', () => {
    result.find(Toggle).simulate('toggle');
    expect(toggleTempAvailability).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('should reset filter when reset button is clicked', () => {
    result.find('#reset-filterDialog').simulate('click');
    expect(resetTempFilters).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });
});
