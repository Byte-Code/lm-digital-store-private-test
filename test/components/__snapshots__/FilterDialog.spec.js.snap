// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDialog should change state when a filter is clicked 1`] = `
<styled.div>
  <styled.div>
    <styled.div
      fFamily="LeroyMerlinSans Light"
      fSize="20px"
      id="close-filterDialog"
      onClick={[Function]}
    >
      <ContentRemoveCircleOutline
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Chiudi Filtri
      </p>
    </styled.div>
    <styled.div
      fSize="16px"
      id="reset-filterDialog"
      onClick={[Function]}
      tDeco="underline"
    >
      <ContentUndo
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Reset Filtri
      </p>
    </styled.div>
  </styled.div>
  <styled.div>
    <p>
      Disponibilità in negozio
    </p>
    <Toggle
      defaultToggled={false}
      disabled={false}
      label="Indifferente"
      labelPosition="right"
      labelStyle={
        Object {
          "alignItems": "center",
          "color": "#fff",
          "display": "flex",
          "fontFamily": "LeroyMerlinSans Italic",
          "fontSize": 16,
          "marginLeft": 50,
        }
      }
      onToggle={[Function]}
      thumbStyle={
        Object {
          "backgroundColor": "#e4e4e4",
          "height": 30,
          "left": 10,
          "top": 10,
          "width": 30,
        }
      }
      thumbSwitchedStyle={
        Object {
          "backgroundColor": "#67cb33",
          "left": 65,
          "top": 10,
        }
      }
      toggled={null}
      trackStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
      trackSwitchedStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
    />
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={true}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <Connect(FilterButton)
    filterMap={
      Object {
        "activeAid": "aid1",
        "activeAvailability": null,
        "activeFilters": Array [
          "filter5",
        ],
        "categoryCode": "CAT123",
      }
    }
    onApply={[Function]}
  />
</styled.div>
`;

exports[`FilterDialog should change state when availability is toggled 1`] = `
<styled.div>
  <styled.div>
    <styled.div
      fFamily="LeroyMerlinSans Light"
      fSize="20px"
      id="close-filterDialog"
      onClick={[Function]}
    >
      <ContentRemoveCircleOutline
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Chiudi Filtri
      </p>
    </styled.div>
    <styled.div
      fSize="16px"
      id="reset-filterDialog"
      onClick={[Function]}
      tDeco="underline"
    >
      <ContentUndo
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Reset Filtri
      </p>
    </styled.div>
  </styled.div>
  <styled.div>
    <p>
      Disponibilità in negozio
    </p>
    <Toggle
      defaultToggled={false}
      disabled={false}
      label="Disponibile"
      labelPosition="right"
      labelStyle={
        Object {
          "alignItems": "center",
          "color": "#67cb33",
          "display": "flex",
          "fontFamily": "LeroyMerlinSans Italic",
          "fontSize": 16,
          "marginLeft": 50,
        }
      }
      onToggle={[Function]}
      thumbStyle={
        Object {
          "backgroundColor": "#e4e4e4",
          "height": 30,
          "left": 10,
          "top": 10,
          "width": 30,
        }
      }
      thumbSwitchedStyle={
        Object {
          "backgroundColor": "#67cb33",
          "left": 65,
          "top": 10,
        }
      }
      toggled={true}
      trackStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
      trackSwitchedStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
    />
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={true}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <Connect(FilterButton)
    filterMap={
      Object {
        "activeAid": "aid1",
        "activeAvailability": true,
        "activeFilters": Array [
          "filter5",
        ],
        "categoryCode": "CAT123",
      }
    }
    onApply={[Function]}
  />
</styled.div>
`;

exports[`FilterDialog should initialize status based on filterMap 1`] = `
<styled.div>
  <styled.div>
    <styled.div
      fFamily="LeroyMerlinSans Light"
      fSize="20px"
      id="close-filterDialog"
      onClick={[Function]}
    >
      <ContentRemoveCircleOutline
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Chiudi Filtri
      </p>
    </styled.div>
    <styled.div
      fSize="16px"
      id="reset-filterDialog"
      onClick={[Function]}
      tDeco="underline"
    >
      <ContentUndo
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Reset Filtri
      </p>
    </styled.div>
  </styled.div>
  <styled.div>
    <p>
      Disponibilità in negozio
    </p>
    <Toggle
      defaultToggled={false}
      disabled={false}
      label="Indifferente"
      labelPosition="right"
      labelStyle={
        Object {
          "alignItems": "center",
          "color": "#fff",
          "display": "flex",
          "fontFamily": "LeroyMerlinSans Italic",
          "fontSize": 16,
          "marginLeft": 50,
        }
      }
      onToggle={[Function]}
      thumbStyle={
        Object {
          "backgroundColor": "#e4e4e4",
          "height": 30,
          "left": 10,
          "top": 10,
          "width": 30,
        }
      }
      thumbSwitchedStyle={
        Object {
          "backgroundColor": "#67cb33",
          "left": 65,
          "top": 10,
        }
      }
      toggled={null}
      trackStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
      trackSwitchedStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
    />
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={true}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={true}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <Connect(FilterButton)
    filterMap={
      Object {
        "activeAid": "aid1",
        "activeAvailability": null,
        "activeFilters": Array [
          "filter0",
          "filter5",
        ],
        "categoryCode": "CAT123",
      }
    }
    onApply={[Function]}
  />
</styled.div>
`;

exports[`FilterDialog should reset filter when reset button is clicked 1`] = `
<styled.div>
  <styled.div>
    <styled.div
      fFamily="LeroyMerlinSans Light"
      fSize="20px"
      id="close-filterDialog"
      onClick={[Function]}
    >
      <ContentRemoveCircleOutline
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Chiudi Filtri
      </p>
    </styled.div>
    <styled.div
      fSize="16px"
      id="reset-filterDialog"
      onClick={[Function]}
      tDeco="underline"
    >
      <ContentUndo
        color="#fff"
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
      <p>
        Reset Filtri
      </p>
    </styled.div>
  </styled.div>
  <styled.div>
    <p>
      Disponibilità in negozio
    </p>
    <Toggle
      defaultToggled={false}
      disabled={false}
      label="Indifferente"
      labelPosition="right"
      labelStyle={
        Object {
          "alignItems": "center",
          "color": "#fff",
          "display": "flex",
          "fontFamily": "LeroyMerlinSans Italic",
          "fontSize": 16,
          "marginLeft": 50,
        }
      }
      onToggle={[Function]}
      thumbStyle={
        Object {
          "backgroundColor": "#e4e4e4",
          "height": 30,
          "left": 10,
          "top": 10,
          "width": 30,
        }
      }
      thumbSwitchedStyle={
        Object {
          "backgroundColor": "#67cb33",
          "left": 65,
          "top": 10,
        }
      }
      toggled={false}
      trackStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
      trackSwitchedStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
    />
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div />
      <styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
        <styled.div
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <Connect(FilterButton)
    filterMap={
      Object {
        "activeAid": "aid1",
        "activeAvailability": false,
        "activeFilters": Array [],
        "categoryCode": "CAT123",
      }
    }
    onApply={[Function]}
  />
</styled.div>
`;
